text/0265-event-vision.md

```diff
@@ -0,0 +1,72 @@
+# Event basics
+
+## Motivation
+
+The introduction of Events has been contentious, so we want to write down and agree on a few basics.
+
+### What are events?
+
+Events are a type of log which has a required event name and a specific structure which is implied by that event name.
+
+The event name and its implied structure make events much more suitable for observability compared to traditional logs.
+
+Because of this, the OpenTelemetry project wants to encourage the transition from traditional logs to events.
+
+### OTLP
+
+Since events are a type of log, they share the same OTLP data structure and OTLP pipeline.
+
+### API
+
+OpenTelemetry should have an Event API. This will help to promote the distinction between events and traditional logs,
+and encourage the use of events over traditional logs.
+
+### Interoperability with traditional logs
+
+It should be possible to send events from the Event API to a traditional logging library (e.g. Log4j).
```

Comment:
I personally think this is a terrible idea, which will lead to a tremendous amount of confusion about what events are to be used for. If an event is just a structured log entry, then I think we should dump events altogether, as they add zero value over structured logging.

Reply:
We can't have an event API, encourage its use instead of traditional logging APIs (slf4j, log4j, etc), and not have mechanisms for the data to be printed to traditional places like the console, files, etc. Nobody would use the event API.  

>  If an event is just a structured log entry

Events [are not just structured log entries](https://github.com/open-telemetry/oteps/pull/265#discussion_r1734979381).

Reply:
As I commented above, if events are "logs v2", then they should _not_ be put in the same stream of data as "logs v1", otherwise, there is no difference, and events _are_ simply structured v1 log entries, perhaps with some sort of semantic meaning (that is completely non-obvious once the event has been co-mingled into the v1 log stream).

Reply:
If that is actually the final state of things, then I strongly suggest we don't call them "events" and we just call them "semantically defined structured logs".

Reply:
> If that is actually the final state of things, then I strongly suggest we don't call them "events" and we just call them "semantically defined structured logs".

+1

Reply:
would this proposal imply renaming existing Event API to `SemanticallyDefinedStructuredLogging` (😒 ) ? Or `Logging API` (super-confusing with logs v1) ?

I think we need to draw a distinction in the name since events represent a subset of logs. Not all logs are events.

Reply:
@reyang is your concern about ILogger, C++ logger, etc where we can have first-class support for events as a part of logging and it makes no sense to provide a separate API?

I assume we could have the distinction in the logging API via `Logger.log` and `Logger.event` or anything else (`[Event]` attribute instead of `[LogMessage]` attribute in .NET). 

 What's the perfect story you'd like to have?

Reply:
Having a name for "structured logs with semantic rigor" would help a lot with education / messaging / promoting their importance.

And having an API for these things would give OpenTelemetry first-class API coverage for all types of telemetry that can be described via semantic conventions.

Given the ambiguity of word "event", and given that it's a reasonably close fit, I personally think co-opting it into a specific definition for "OpenTelemetry Event™" seems vaguely reasonable.

Reply:
> What's the perfect story you'd like to have?

https://github.com/open-telemetry/opentelemetry-specification/pull/3167#pullrequestreview-1283987908

I think it'll be great if OpenTelemetry can influence the logging API design across programming languages/runtimes. More details here https://github.com/open-telemetry/opentelemetry-specification/pull/3167#pullrequestreview-1283987908

To give some examples, I think it'll be awesome if we can say something like:

1. Rust has well established logging APIs from https://github.com/tokio-rs/tracing, OpenTelemetry worked closely with the Rust and Tokio communities, these APIs were enhanced to cover all the features.
2. .NET runtime has well established ILogger API, it already supports 95% of the features from what OpenTelemetry would need. .NET Foundation worked closely with OpenTelemetry community to close the 5% feature gap and now it covers 100% of the scenarios and features.
3. C++ doesn't have standard logging API, OpenTelemetry made progress to ship the logging API, and this work eventually influenced the C++ Standard Committee, so we worked together and put the API as part of the C++ standard libraries.

I think it'll be bad if we have something like:

1. SLF4J only covers 80% of what OpenTelemetry needs. OpenTelemetry ended up inventing a new set of competing APIs, now we have to do bridges from both sides and customers have to pick which one to use.

Reply:
> I personally think this is a terrible idea, which will lead to a tremendous amount of confusion about what events are to be used for. If an event is just a structured log entry, then I think we should dump events altogether, as they add zero value over structured logging.

+

> As I commented above, if events are "logs v2", then they should not be put in the same stream of data as "logs v1". Once the data is in the log stream, there is effectively no difference between v1 and v2 (aside from defined semantic attributes)

I totally agree. If we are trying to promote a new thing, which we believe is better than traditional logs for observability, why not just have this concept in the SDK as done for metrics/spans? We just happen to use the same exporter and share the same data model but that's it. 

> We can't have an event API, encourage its use instead of traditional logging APIs (slf4j, log4j, etc), and not have mechanisms for the data to be printed to traditional places like the console, files, etc. Nobody would use the event API.

I may be missing something, but I don't see this as a strong argument. If I'm a end user, and want to use OTel events and I understand they are not logs, why would I expect it to be in places where normal logs are? 

We maybe be better off stop relating events with logs. Just because they are transmitted the same way via OTLP, they are not the same.

Reply:
> I think it'll be great if OpenTelemetry can influence the logging API design across programming languages/runtimes.

@reyang the problem I see with this is that we don't just want to enhance existing logging APIs designs to have coverage for the OpenTelemetry event concept. We also want to actively discourage the use of traditional logs. An SLF4J logger which allows you to produce OpenTelemetry events still has 80%+ API surface area dedicated to producing traditional logs. Its too easy to do the wrong thing. The story would be better if SLF4J added dedicated support for OpenTelemetry events AND deprecated everything else, but that's not realistic.

Even if SLF4J were to have a perfect OpenTelemetry event story, we still would be stuck with the downside of having an inconsistent API story: "to use OpenTelemetry, take a dependency on the OpenTelemetry metrics and trace API, and the SLF4j logging / event API". Not necessarily a deal breaker, but it is a downside. Its cleaner to say "to Use OpenTelemetry, take a dependency on the OpenTelemetry API (for tracing, metrics, and events)". 

>  If we are trying to promote a new thing, which we believe is better than traditional logs for observability, why not just have this concept in the SDK as done for metrics/spans? We just happen to use the same exporter and share the same data model but that's it.

@joaopgrassi I think @jkwatson is saying that events should be a new proto type, so we'd have tracing, metrics, logs, and events. The reason we shouldn't do this is because the data model / proto message structure for events and logs would be essentially identical.

> If I'm a end user, and want to use OTel events and I understand they are not logs, why would I expect it to be in places where normal logs are? I use the API to produce events, I configure my OTLP exporter, I get the logs in my back-end and that's it.

We want library authors to use opentelemetry APIs for all observability. This means choosing the event API over existing log APIs. Library authors won't make this decision if they view opentelemetry APIs as overly restrictive about where the data ends up. From an opentelemetry standpoint, we expect data to be sent to a network location over OTLP, but for the large majority of users for which opentelemetry is still a new concept, they'll continue to expect to be able to turn on debug logging for a library and see the data on the console. 

I feel quite strongly about this: **library authors will not elect to replace logging APIs with the event API unless events can be _easily_ routed to typical log frameworks**.

Reply:
> customers have to pick which one to use

this OTEP says that "OpenTelemetry will recommend that application developers also use the Event API over emitting events via a generic logging library".

Reply:
In the perfect world, I'd love us to go work with runtimes, languages and popular logging facades and implement 'better logging' there. 

We will probably succeed in some places and fail in others. Problems:

1. It'll take years or decades or have proper event support in most languages and popular libs
    
    E.g. in Java, you'd find at least 3 logging APIs that people use:
    - Java Util Logging (built-in into the language, does not support any structure)
    - SLF4J (supports structure prioritizing string messages)
    - Log4J API (supports structure prioritizing string messages)
    
    They all are extremely conservative and despite existing for two+ decades, they at major version 2. (log4j started doing betas for v3 recently that improves support for structured logs even more). 
    Despite releasing v2 major version recently (with structured logs support), SLF4J didn't break any public APIs and still prioritizes string messages. 

2. In languages where multiple logging libs are popular, OTel will need to pick one to depend on
    Assuming we don't provide better-than-logs API, OTel instrumentation libraries that will exist for years will need to pick one of the external logging libraries to emit events. 

3. We'll need to develop all sorts of weird conventions on how to emit events with every logging lib out there. It'll be hard and error prone

Reply:
WRT proto/SDK:

If we could go back in time and could name `LogRecord` an `Event` instead (not possible), things could be the following:

- Event API and SDK encourage structured semantical events. Event proto represents them on OTLP
- 3rd party logging API usually creates string-body with optional attributes and uses Event API as a facade. (we still need a way to forward Event API to traditional logging systems and should be cautious of loops)

I.e. logs API is supported on top of events, logging systems/files/stdout can be used as exporting destinations, but things in the middle are the same.  What's the benefit of separating them (and asking users to understand subtle differences, configure two pipelines, etc)?

Reply:
If we try to define a new logging API for Java, I can guarantee with near 100% certainty, that it will never be used by anyone outside the community. Java has multiple logging APIs already, and convincing people to switch would be a futile task. That doesn't even touch on the issue that having a first class logging API will require having a large configuration surface area, including formatters, appenders, etc etc.

Reply:
That's why we're not trying to introduce a new logging API, we're using Event API which is already there and is experimental in Java:
 - it does not need formatters since it does not need string formatting
 - it's configuration is OTel configuration (do we want config to disable tracer by name? now we'll add config that configures `EventLogger` by name and verbosity)
 - OTel appenders are exporters and we only need a way to forward Event API to traditional logging appenders

Reply:
I think this is an unrealistic expectation. If we treat this as a type of logging, people are going to ask for formatters and appenders, and advanced configuration, to match what they are used to from existing logging libraries.

Reply:
> this is definitely an option, to frame Events as a new type (as opposed to a subset of Logs) which just happen to share a transport mechanism with Logs

Seems we are going in circles, and I wonder why we haven't discussed this option. Events are a new thing, why do we want to consider it as a log which carries ages of expectations and existing features on its back. I feel this could be much simpler if we just say hey, OTel has events, here's the API how to use it.

Reply:
> If we treat this as a type of logging, people are going to ask for formatters and appenders, and advanced configuration, to match what they are used to from existing logging libraries.

Having an Event API and saying that the Events are a type of logging doesn't require formatters appenders. We can accommodate this by providing capabilities to route Event API calls back to existing log frameworks and their rich ecosystems of formatters / appenders. This could be done via _an alternative_ implementation of the Event API. I.e. the user installs in a Logback / Log4j / JUL implementation of the OpenTelemetry Event API instead of the OpenTelemetry SDK. OR, the OpenTelemetry SDK routes events back to Logback / Log4j / JUL in addition to exporting them to a network location via OTLP.

Reply:
I think we have 3 discussions in this thread:

1. are Event API and traditional logging API the same? 
    - I think no - events are very opinionated logs (structure is forced) https://github.com/open-telemetry/oteps/pull/265#issuecomment-2322236851
2. do we need proper structured logging API in otel 
    - based on today's spec discussion - probably yes
    - would this API be any different from events API??
3. do we need separate SDK/pipeline for events and logs
    - ?

Reply:
I'm going to write down the anology on (2) Do we need a proper structured logging API in otel that I gave in the spec meeting.

I think the answer to this question is *yes*. I also agree with @jkwatson that we don't want to force users to give up existing logging solutions they use today.  However - I disagree that we should use this as a justification for not providing an API for users.

The analogy (borrowed from someone else) is a Ski resort.  Ski resorts have ski rental shops.   Many folks who Ski bring their own skis, or may rent skis from other rental shops. The Ski resort still has its own rental shop because some folks will need to rent skis.   This is because the goal of the Ski resort is for folks to Ski.  Rentals may be a necessary component, but the Ski resort doesn't really care if folks rent somewhere else or bring their own Skis.

OpenTelemetry is about correlated, causal telemetry.  We need to interoperate with Logs (Skis, if you will) no matter how people bring them.  We should also provide our own API, given the demand from users and the need for us to be a complete solution.  However, we need *first class* support for existing logging APIs.

This is the reality of OpenTelemetry and *any existing, well established o11y API in any ecosystem*.

Reply:
@jsuereth Using this analogy, would our Ski resort be interested in encouraging specific types of equipment we rent/allow to use? I think we need to do it. 

It's totally valid to use 3rd party skis created 30 years ago, but our lifts and slopes are optimized for different use-case. We won't sell or rent skis based on 30 years old designs in our shops.

To get back to logs, our potential logging API should prioritize structure and discourage string formatting. We won't be able to provide great experience for all 3rd party logs out there. We don't need to support string formatting features since it's bad for correlation.

If we want to have structured logs API, I believe it should be the same API as events. Overloads that take event name produce named-structured-logs (aka events), overloads that don't take name produce regular structured logs.

Reply:
@lmolkova Yes - I'd agree with that as well.

Reply:
> If we want to have structured logs API, I believe it should be the same API as events. Overloads that take event name produce named-structured-logs (aka events), overloads that don't take name produce regular structured logs.

I see this as a distinction without difference. An event emitted through the event API shouldn't cause an error, it should just emit an untyped event, aka, a structured log.

Reply:
"All Events are structured logs, not all structured logs are Events" but the Event API can create either.

Reply:
>An event emitted through the event API shouldn't cause an error, it should just emit an untyped event, aka, a structured log.

did you mean a **log** emitted through event API ? 

If so, I agree with this, we should have a general-purpose structured logging API capable of emitting named structured logs which might be called Events or something else https://github.com/open-telemetry/oteps/pull/265#issuecomment-2329633446

Reply:
I meant an event in the generic sense, which could also be called a log, yes.

events = things that happen
Events = events with a defined semantic meaning
Untyped Events = structured events with no defined semantic meaning?

I think it's more important we figure out the taxonomy of how we want things to work in otel vs. how they comport with other systems. Obviously you can have structured logs in log4j. You can even have your own internal schema for those structured logs, making them logically equivalent to OTel Events. The differences are the things I noted in https://github.com/open-telemetry/oteps/pull/265#issuecomment-2327290693; namely, context.

Reply:
Respectfully, but this whole debate – do need to make logs? And if so, do we need an API? And if so, what do we call the API? I don't think this is where the conversation really is. Over the past two years we've laid those questions to rest. If there are objectors, I get it, naming is hard. But I don't want us unravelling those decisions. The point of this doc is not to throw them out, it is to try and explain them properly. 

That said, here are some additional details that can hopefully explain why we have an Event API but no Event data structure.

### Why have an event API?

Otel has two goals when it comes to logging: 

* Consume all of the end users' existing logs.
* Produce our own highly opinionated logs, which will replace span events.

These goals are at the opposite ends of the spectrum. When consuming existing logs we want to be as flexible as possible, when creating our own logs we want to be as rigorous as possible. That's why we have both the `LogBridge` API and the `Event` API.

We can't avoid having our own instrumentation API. We have a very specific concept, and we need an API that allows us to express it correctly. Using the general purpose LogBridge API does not suffice, and we can't take a dependency on third party logging frameworks.

### Why aren't events a separate signal with their own data structure?

OpenTelemetry already has two data structures for events: LogRecords and SpanEvents. These two separate data structures exist because there are two kinds of backends: tracing backends and logging backends. 

Unified backends are relatively new.  We expect these new unified backends to store all events as LogRecords, and to deprecate the SpanEvent data structure. 

The point is: data structures are a configuration issue. Some end users will want the events to be recorded as LogRecords. Other end users will want events to be recorded as SpanEvents. Others will want both. It completely depends on their setup.

### Should we call these logs or events?

Otel events are both structured logs and the successor to SpanEvents. So what should we call them?

We choose the name Events as our term for Otel structured logging because it avoids the most confusion. There are many kinds of "structured logging" already involved with both the LogBridge API and the Collector logging pipeline. The LogRecord data structure itself can be thought of as "Otel structured logging."  

On the other hand, the term "event" is less confusing. Not because there aren't many theoretical definitions for the meaning of event, but because there aren't many practical "event" systems that users are trying to integrate with. By referring to our specific structure as an event helps to avoid confusion with the other forms of structured logging that are present in every user's logging pipeline.

Reply:
trying to summarize here and bring these threads together into a decision: https://github.com/open-telemetry/oteps/pull/265#issuecomment-2334806152

Reply:
Resolving this based on latest updates. If you'd like to continue discussing, please start a new thread with a summary of any remaining points you'd like to discuss. Thanks.